neps/nep-0621.md

```diff
@@ -0,0 +1,490 @@
+    fn withdraw(&mut self, assets: U128, receiver_id: Option<AccountId>) -> PromiseOrValue<U128>;
+
+    /// Simulates withdrawing exactly `assets` worth of tokens without executing.
+    ///
+    /// Differs from [`Self::convert_to_shares`] by factoring in:
+    /// - The caller's current share balance.
+    /// - Vault withdrawal limits.
+    /// - Applicable fees or penalties.
+    ///
+    /// # Panics / Fails
+    /// - The `preview_withdraw` method must never panic.
+    ///   It should return the number of shares required to withdraw the given amount of assets,
+    ///   or `0` if withdrawals are not currently allowed.
+    ///
+    /// Useful for frontends to preview required shares for a given withdrawal.
+    fn preview_withdraw(&self, assets: U128) -> U128;
+}
+```
+
+### Deposit and Mint
+
+```rust
+pub trait FungibleTokenReceiver {
+    /// # Events
+    /// - The vault **must** emit a `VaultDeposit` event upon a successful deposit.
+    /// - The vault should also emit a `FtMint` event for the share they issued.
+    fn ft_on_transfer(&mut self, sender_id: AccountId, amount: U128, msg: String) -> PromiseOrValue<U128>
+}
+```
+
+-   According to the NEP-141 standard, all fungible token transfers to a contract must invoke the `ft_on_transfer` callback.
+-   In the context of ERC-4626, both the **`deposit`** and **`mint`** functions are implemented through this entrypoint on the vault contract.
+-   While NEP-141 does not prescribe how the `msg` parameter should be structured, our suggested convention is to pass a JSON-encoded message with the following schema:
+
+```rust
+pub struct DepositMessage {
+    /// The minimum number of shares that must be received for this deposit to succeed. (Slippage control)
+    min_shares: Option<U128>,
+    /// The maximum number of shares that can be minted, used for `mint` operations.
+    max_shares: Option<U128>,
+    /// The account that should receive the minted shares.
+    receiver_id: Option<AccountId>,
+    /// Optional memo for logging or off-chain indexing.
+    memo: Option<String>,
+    /// If `true`, the transfer is treated as a donation and no shares are minted.
+    donate: Option<bool>,
+}
+```
+
+-   Implementers are free to define their own schema for `msg`, but the vault must correctly handle `ft_on_transfer` according to NEP-141:
+    -   Return `PromiseOrValue::Value(0)` if all tokens were accepted.
+    -   Return a non-zero `U128` to indicate the number of tokens to refund.
+
+### Optional Slippage Control:
+
+-   To mitigate the risk of unfavorable execution due to block delays, vault contracts may optionally implement the following methods, which act as "slippage-protected" counterparts to `redeem` and `withdraw`.
+
+```rust
+pub trait VaultOptionalSlippageWithdraw {
+    /// Redeem a specific number of shares, **only if** they are worth
+    /// at least `min_amount` of assets.
+    /// This acts as a slippage-protected version of `redeem`.
+    fn redeem_with_slippage(
```

Comment:
It seems both `redeem` and `withdraw` must have slippage control via `min_amount` parameter by default on NEAR, without "optional for security" `redeem/withdraw_with_slippage` counterparts. Because NEAR is fundamentally asynchronous in execution, and because this NEP is adaptation to NEAR's reality, it's worth to introduce secure by default methods, not vice versa.